File: day10/day10.mjs
import readDataFromStdin from "../read-data.mjs";
const data = await readDataFromStdin()

let x = 1;
const signals = []

data.split("\n").forEach(instruction => {
    const [command, arg] = instruction.split(" ");
    if (command === "noop") {
        signals.push(x);
    } else if (command === "addx") {
        signals.push(x);
        signals.push(x);
        x += Number(arg);
    } else {
        throw new Error(`Unknown command: ${command}`);
    }
});

const freqs = [20, 60, 100, 140, 180, 220]
const strengths = freqs.map(freq => signals[freq - 1] * freq)
const part1 = strengths.reduce((a, b) => a + b, 0)

console.log({ x, strengths })
console.log("Part 1:", part1)

// signals.forEach((level, index) => {
//     if (index < 10) {
//         console.log(index, level);
//     }
// })
signals.forEach((level, index) => {
    const calcIndex = index % 40;
    if (Math.abs(level - calcIndex) <= 1) {
        process.stdout.write("#");
    } else {
        process.stdout.write(".");
    }
    if ((index + 1) % 40 === 0) {
        process.stdout.write("\n");
    }
});

File: day10/day10.exs
input =
  IO.read(:stdio, :all)
  |> String.trim()
  |> String.split("\n")

cycles =
  input
  |> Enum.reduce([], fn line, acc ->
    level = (List.last(acc) || [1, 1]) |> Enum.at(0)

    if line == "noop" do
      acc ++ [[level, level]]
    else
      num = line |> String.split(" ") |> Enum.at(1) |> String.to_integer()
      acc ++ [[level, level], [level + num, level]]
    end
  end)

[20, 60, 100, 140, 180, 220]
|> Enum.map(fn f ->
  l = Enum.at(cycles, f - 1) |> Enum.at(1)
  l * f
end)
|> Enum.sum()
|> IO.inspect(label: "Part 1")

IO.puts("Part 2:")

cycles
|> Enum.with_index()
|> Enum.each(fn {[_, level], index} ->
  if abs(level - rem(index, 40)) <= 1 do
    IO.write("#")
  else
    IO.write(".")
  end

  if rem(index + 1, 40) == 0 do
    IO.puts("")
  end
end)
